src/hades/leases/server.py

```diff
@@ -0,0 +1,515 @@
+            stderr = ensure_stream_writable(streams[2], 'stderr')
+            # Clear the stack
+            stack.pop_all()
+            return (stdin, stdout, stderr), argv, environ
+
+    @staticmethod
+    def parse_ancillary_data(
+            ancdata: Container[Tuple[int, int, bytes]],
+            expected_fd_modes: Sequence[str],
+    ) -> List[TextIO]:
+        """
+        Open streams for file descriptors received via :func:`socket.recvmsg`
+        ancillary data.
+
+        :param ancdata:
+        :param expected_fd_modes: a sequence of modes in which the fds should be opened
+        :return:
+        """
+        fds = array.array("i")
+        streams = []
+        with contextlib.ExitStack() as stack:
+            truncated = False
+            for cmsg_level, cmsg_type, cmsg_data in ancdata:
+                if (
+                        cmsg_level == socket.SOL_SOCKET
+                        and cmsg_type == socket.SCM_RIGHTS
+                ):
+                    end = len(cmsg_data) - (len(cmsg_data) % fds.itemsize)
+                    truncated |= end != len(cmsg_data)
+                    fds.frombytes(cmsg_data[:end])
+                else:
+                    logger.warning(
+                        "Received unexpected control message: level=%d type=%d",
+                        cmsg_level, cmsg_type,
+                    )
+            # Ensure that file descriptors get closed on error
+            for fd in fds:
+                stack.callback(_try_close, fd)
+            if truncated:
+                raise ProtocolError(
+                    "Received truncated file descriptor. "
+                    "SCM_RIGHTS control message data must be an multiple of "
+                    "sizeof(int) = {}".format(fds.itemsize)
+                )
+            for fd, fd_mode in zip_left(fds, expected_fd_modes):
+                flags = fcntl.fcntl(fd, fcntl.F_GETFL)
+                if flags & os.O_ACCMODE == os.O_RDONLY:
+                    mode = "r"
+                elif flags & os.O_ACCMODE == os.O_WRONLY:
+                    mode = "w"
+                elif flags & os.O_ACCMODE == os.O_RDWR:
+                    # the stdout/stderr buffers can possibly be in RW mode,
+                    # however the buffer used by `sys.stdout` is usually opened in
+                    # write-only mode by python.
+                    # in fact, opening this in `r+` (i.e. read-write mode) and using buffering
+                    # causes open() to refuse operation because the buffer is not seekable.
+                    # See https://bugs.python.org/issue20074#msg207012 and the related discussion
+                    # for some details on the core developers' philosophy on this.
+                    mode = fd_mode or "w"
+                else:
+                    os.close(fd)
+                    logger.warning("Unknown fd ACCMODE %x for fd %d", flags & os.O_ACCMODE, fd)
+                    continue
```

Comment:
I think `continue`ing here is a bug, as it messes file descriptors role associations, e.g. someone sends four fds with the first one having an illegal `O_ACCMODE`, then we will happily map the second fd to `stdin` etc. I think, we should `raise ProtocolError` here.